Restructure import assertions description to ease maintainability (#42309)
mdn/content: files/en-us/web/javascript/reference/statements/import/with/index.md

- `"dataN"`
  - : An attribute value. Must be a string literal.

### Exceptions

- {{jsxref("SyntaxError")}}
  - : An unsupported `key` was specified.

- {{jsxref("TypeError")}}

>   - : An unsupported value was used with a supported `key`.

Reviewer: Sooo, this is complicated. `import foo from "foo" with { unlikely: "bah" }` is a SyntaxError, but `import("foo", { with: { unlikely: "bah" } })` is a TypeError. As for unsupported values, the host is not _required_ to throw an error, or rather it can choose to do anything _including_ throwing an error, and the behavior is wholly dependent on the `key` itself. This was chiefly why I left it off because basically nothing is certain, but okay if you want to add it; just need to point out these two things.

{"name":"Maria"}
```

Modules are identified and parsed only according to their served [media type (MIME type)](/en-US/docs/Web/HTTP/Guides/MIME_types) — the file extension in the URL cannot be used to identify a file's type. In this case, the MIME type is `application/json`, which tells the browser that the file is JSON and must be parsed as JSON. If, for some reason (e.g., the server is hijacked or bogus), the MIME type in the server response is set to `text/javascript` (for JavaScript source), then the file would be parsed and executed as code. If the "JSON" file actually contains malicious code, the `import` declaration would unintentionally execute external code, posing a serious security threat.

Import attributes fix this problem by allowing the author to explicitly specify how a module should be validated.

> When using import attributes the code above would be written as shown below, and would fail if the file was served with any media type other than `application/json`:

Author: Thanks. I've attempted to more closely match your intent in https://github.com/mdn/content/pull/42309/changes/3300ed07ea22a33ab969e3da1794f066a3204cd7

```js
import data from "https://example.com/data.json" with { type: "json" };
```

> A runtime can read and process any attributes that it understands.

Reviewer: This sounds tautological. Perhaps just "all attributes you specify must be understood by the runtime, otherwise a syntax error is thrown"?

Author: Yes, that's better

```suggestion
All attributes you specify must be understood by the runtime, otherwise a syntax error is thrown.
```

However, you cannot use unknown attributes — a runtime will throw a syntax error if it encounters an unknown attribute.

### Standard attributes

> The available attributes depend on the language and runtime environment (the ECMAScript standard does not define any attributes).

Reviewer: This isn't actually true. ECMAScript defines `type: "json"`; see https://tc39.es/ecma262/multipage/ecmascript-language-scripts-and-modules.html#sec-HostLoadImportedModule

Author: Thanks! Fixed in https://github.com/mdn/content/pull/42309/commits/600f8b6fd13fe2ddfafa9523fde717acbe385f2b

Browsers, on the other hand, implicitly assume that the module is JavaScript if the `type` is not specified, and fail if the module is not JavaScript (for example, JSON). This ensures that module types are always strictly validated and prevents any security risks.
Non-browser runtimes such as Node and Deno align with browser semantics and enforce `type` for JSON modules.

In other words, if you omit the `type` and attempt to import a file as `"application/json"` you will usually get an error like the following:

```plain
Failed to load module script: Expected a JavaScript module script but the server responded with a MIME type of "application/json". Strict MIME type checking is enforced for module scripts per HTML spec.
```

#### CSS Modules (`{ type: "css" }`)

The HTML spec defines the `css` type, which imports a stylesheet into a script as an {{domxref("CSSStyleSheet")}} object.

The code below shows how you might import a style and add it to your document.
The import will throw an exception if `example_styles.css` is served with any media type other than `"text/css"`.

```js
import exampleStyles from "https://example.com/example_styles.css" with { type: "css" };

> document.adoptedStyleSheets = [exampleStyles];

Reviewer: ```suggestion

document.adoptedStyleSheets.push(exampleStyles);
```

```

Note that importing CSS modules into workers is not usually supported.

> It is allowed by the HTML specification if `CSSStyleSheet` is exposed to workers, but the CSSOM specification only exposes `CSSStyleSheet` in the main window thread.

Author: ```suggestion
Note that importing CSS modules into workers is usually not supported, because  the CSSOM specification only exposes `CSSStyleSheet` in the window context.
```
